volumina/utility/preferences.py

```diff
@@ -0,0 +1,143 @@
+# as published by the Free Software Foundation; either version 2.1
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU Lesser General Public License for more details.
+#
+# See the files LICENSE.lgpl2 and LICENSE.lgpl3 for full text of the
+# GNU Lesser General Public License version 2.1 and 3 respectively.
+# This information is also available on the ilastik web site at:
+# 		   http://ilastik.org/license/
+###############################################################################
+
+"""Application-wide persistent preferences.
+
+All public functions in this module are thread-safe.
+"""
+
+import logging
+import os
+import pathlib
+import pickle
+import threading
+from typing import Any, Tuple, MutableMapping
+
+logger = logging.getLogger(__name__)
+
+
+class _Preferences:
+    def __init__(self, path: pathlib.Path):
+        self._path = path
+        self._data = _load_preferences(path)
+        self._lock = threading.Lock()
+
+    def get(self, group: str, setting: str, default: Any = None) -> Any:
+        return self.getmany((group, setting, default))[0]
+
+    def set(self, group: str, setting: str, value: Any) -> None:
+        self.setmany((group, setting, value))
+
+    def getmany(self, *args: Tuple[str, str, Any]) -> Tuple[Any, ...]:
+        with self._lock:
+            result = []
+            for group, setting, default in args:
+                try:
+                    result.append(self._data[group][setting])
+                except KeyError:
+                    result.append(default)
+            return tuple(result)
+
+    def setmany(self, *args: Tuple[str, str, Any]) -> None:
+        with self._lock:
+            for group, setting, value in args:
+                self._data.setdefault(group, {})[setting] = value
+            try:
+                with open(self._path, "wb") as f:
+                    # Use the lowest available protocol for backwards compatibility.
+                    pickle.dump(self._data, f, protocol=0)
+            except Exception as e:
+                logger.exception(f"Failed to save preferences to {str(self._path)!r}")
+                if not isinstance(e, IOError):
+                    self._path.unlink()
+
+    def set_location(self, path: pathlib.Path) -> None:
+        with self._lock:
+            self._path = path
+            self._data = _load_preferences(path)
+
+
+def _load_preferences(path: pathlib.Path) -> MutableMapping[str, MutableMapping[str, Any]]:
+    try:
+        with open(path, "rb") as f:
+            return pickle.load(f)
+    except FileNotFoundError:
+        return {}
+    except Exception as e:
+        logger.exception(f"Failed to load preferences from {str(path)!r}")
+        if not isinstance(e, IOError):
```

Comment:
Should we clarify why we are doing this in the comment and why we are checking for IOError?
Also if the parent folder of this file has only read permissions unlink will raise PermissionDenied.

Reply:
`IOError` means that we (probably) won't be able to unlink the file, hence the instance check.